pkg/workceptor/kubernetes.go

```diff
@@ -263,182 +273,326 @@ func (kw *kubeUnit) createPod(env map[string]string) error {
+						podNamespace,
+						podName,
+						err,
+					)
+					time.Sleep(100 * time.Millisecond)
 				} else {
 					break
 				}
 			}
-			if errStdin != nil {
-				logStream.Close()
+
+			if err != nil {
+				stdinErr = err
+				errMsg := fmt.Sprintf("Error streaming stdin to pod %s/%s: %s",
+					podNamespace,
+					podName,
+					err,
+				)
+				kw.Error(errMsg)
+				kw.UpdateBasicStatus(WorkStateFailed, errMsg, stdout.Size())
+
+				close(stdinErrChan) // signal STDOUT goroutine to stop
+			} else {
+				if stdin.Error() == io.EOF {
+					kw.UpdateBasicStatus(WorkStateRunning, "Pod Running", stdout.Size())
+				} else {
+					// this is probably not possible...
+					errMsg := fmt.Sprintf("Error reading stdin: %s", stdin.Error())
+					kw.Error(errMsg)
+					kw.UpdateBasicStatus(WorkStateFailed, errMsg, stdout.Size())
+
+					close(stdinErrChan) // signal STDOUT goroutine to stop
+				}
 			}
-			streamWait.Done()
 		}()
 	}
+
+	// TODO: @seth continue and go through this goroutine
 	go func() {
-		_, errStdout = io.Copy(stdout, logStream)
-		streamWait.Done()
+		defer streamWait.Done()
+
+		var sinceTime time.Time
+		var logStream io.ReadCloser
+		numEOF := 0 // resets to 0 on each successful read from pod stdout
+
+		for {
+			if stdinErr != nil {
+				break
+			}
+
+			// get pod, with retry
+			for retries := 5; retries > 0; retries-- {
+				kw.pod, err = kw.clientset.CoreV1().Pods(podNamespace).Get(kw.ctx, podName, metav1.GetOptions{})
+				if err == nil {
+					break
+				} else {
+					time.Sleep(100 * time.Millisecond)
+				}
+			}
+			if err != nil {
+				errMsg := fmt.Sprintf("Error getting pod %s/%s: %s", podNamespace, podName, err)
+				kw.UpdateBasicStatus(WorkStateFailed, errMsg, 0)
+				kw.Error(errMsg)
+
+				break
+			}
+
+			logReq := kw.clientset.CoreV1().Pods(podNamespace).GetLogs(
+				podName, &corev1.PodLogOptions{
+					Container:  "worker",
+					Follow:     true,
+					Timestamps: true,
+					SinceTime:  &metav1.Time{Time: sinceTime},
+				},
+			)
+			// get logstream, with retry
+			for retries := 5; retries > 0; retries-- {
+				logStream, err = logReq.Stream(kw.ctx)
+				if err == nil {
+					break
+				} else {
+					time.Sleep(100 * time.Millisecond)
+				}
+			}
+			if err != nil {
+				errMsg := fmt.Sprintf("Error opening pod %s/%s stream: %s", podNamespace, podName, err)
+				kw.UpdateBasicStatus(WorkStateFailed, errMsg, 0)
+				kw.Error(errMsg)
+
+				break
+			}
+
+			// read from logstream
+			streamReader := bufio.NewReader(logStream)
+			for stdinErr == nil { // check between every line read to see if we need to stop reading
+				line, err := streamReader.ReadString('\n')
+				if err == io.EOF {
+					kw.Debug("detected EOF for pod %s/%s, attempt %d", podNamespace, podName, numEOF)
+					numEOF++
+					if numEOF <= 5 {
+						time.Sleep(100 * time.Millisecond)
+
+						break
+					}
+
+					return
+				}
+				if err != nil {
+					stdoutErr = err
+
+					break
```

Comment:
we don't check for stdoutErr anywhere in the outer loop -- this could result in an infinite loop. maybe we need to return here

Reply:
damn good catch